File: python3/DFS,BFS/(programmers)아이템 줍기.py
import numpy as np
from collections import deque
"""
    1. 정방행렬 관련 문제는 numpy를 이용하여 쉽게 해결할 수 있다. 
    2. 이 문제같은 경우 모서리를 탐색해야 하기 때문에 배열을 2배 키워줘야 겹치는 모서리가 발생하지 않는다. 
"""
def solution(rectangle, characterX, characterY, itemX, itemY):
    answer = 0
    characterX*=2
    characterY*=2
    itemX*=2
    itemY*=2
    
    N=51*2
    ground = np.zeros((N,N))
    for i in rectangle:
        ground[2*i[0]:2*i[2]+1, 2*i[1]:2*i[3]+1] = 1
    for i in rectangle:
        ground[2*i[0]+1:2*i[2], 2*i[1]+1:2*i[3]] = 0
    q = deque()
    q.append([characterX, characterY])
    ground[characterX,characterY]=0
    dx=[1,-1,0,0]
    dy=[0,0,1,-1]
    
    while len(q) != 0:
        px, py = q.popleft()
        if px == itemX and py==itemY:
            answer = ground[px][py]//2
            break
        for x, y in zip(dx, dy):
            nx = px+x
            ny = py+y
            if nx>=0 and ny>=0 and nx<N and ny<N and ground[nx,ny]==1:            
                q.append([nx,ny])
                ground[nx,ny]=ground[px,py]+1
            
    return answer

File: c++/BFS/(programmers)아이템 줍기.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>
using namespace std;

int solution(vector<vector<int>> rectangle, int characterX, int characterY, int itemX, int itemY) {
    int answer = 1000000000;
    int N = 51;
    vector<vector<int>> map(N * 2, vector<int>(N * 2, 0));
    vector<vector<int>> visited(map.begin(), map.end());
    for (int i = 0; i < rectangle.size(); i++) {
        int x1 = rectangle[i][0] * 2;
        int y1 = rectangle[i][1] * 2;
        int x2 = rectangle[i][2] * 2;
        int y2 = rectangle[i][3] * 2;
        for (int i = x1; i <= x2; i++) {
            for (int j = y1; j <= y2; j++) {
                map[i][j] = 1;
            }
        }
    }
    for (int i = 0; i < rectangle.size(); i++) {
        int x1 = rectangle[i][0] * 2;
        int y1 = rectangle[i][1] * 2;
        int x2 = rectangle[i][2] * 2;
        int y2 = rectangle[i][3] * 2;
        for (int i = x1 + 1; i < x2; i++) {
            for (int j = y1 + 1; j < y2; j++) {
                map[i][j] = 0;
            }
        }
    }
    int dx[4] = { 0,0,1,-1 }, dy[4] = { 1,-1,0,0 };
    queue<pair<int, pair<int, int>>> q;
    q.push({ 0, {2 * characterX, 2 * characterY} });
    visited[2 * characterX][2 * characterY] = 1;
    while (!q.empty()) {
        int dist = q.front().first;
        int px = q.front().second.first;
        int py = q.front().second.second;
        q.pop();
        if (px == 2 * itemX && py == 2 * itemY) {
            answer = dist / 2;
        }
        for (int i = 0; i < 4; i++) {
            int nx = px + dx[i];
            int ny = py + dy[i];
            if (nx < 0 || nx >= 2 * N || ny < 0 || ny >= 2 * N || visited[nx][ny] == 1 || map[nx][ny] == 0) continue;
            visited[nx][ny] = 1;
            q.push({ dist + 1,{nx,ny} });
        }
    }
    return answer;
}